daemon/containerd/image_manifest.go

```diff
@@ -0,0 +1,146 @@
+package containerd
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/images"
+	containerdimages "github.com/containerd/containerd/images"
+	cplatforms "github.com/containerd/containerd/platforms"
+	"github.com/docker/docker/errdefs"
+	"github.com/moby/buildkit/util/attestation"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+// walkImageManifests calls the handler for each locally present manifest in
+// the image. The image implements the containerd.Image interface, but all
+// operations act on the specific manifest instead of the index.
+func (i *ImageService) walkImageManifests(ctx context.Context, img containerdimages.Image, handler func(img *ImageManifest) error) error {
+	desc := img.Target
+
+	handleManifest := func(ctx context.Context, d ocispec.Descriptor) error {
+		platformImg, err := i.NewImageManifest(ctx, img, d)
+		if err != nil {
+			return err
+		}
+		return handler(platformImg)
+	}
+
+	if containerdimages.IsIndexType(desc.MediaType) {
+		store := i.client.ContentStore()
+		return containerdimages.Walk(ctx, presentChildrenHandler(store, containerdimages.HandlerFunc(
+			func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+				if containerdimages.IsManifestType(desc.MediaType) {
+					return nil, handleManifest(ctx, desc)
+				}
+				return nil, nil
+			})), desc)
+	}
```

Comment:
```suggestion
	if containerdimages.IsIndexType(desc.MediaType) {
		return walkPresentChildren(ctx, desc, handleManifest)
	}
```

Does this simplification work, or do we need to be sure to propagate errors up out of `handler()`? If that is the case, maybe we can modify these helpers to be suitable for this use-case as well?

Reply:
@vvoland ptal 🤗

Reply:
We need to propagate the error. But then, if we change `walkPresentChildren` to handle the error, this will always return an error because not all children of an index are a manifest. To workaround this we would have to ignore the non-manifests in `handleManifests` and reintroduce the check from https://github.com/moby/moby/pull/45644#discussion_r1210179649

Pushed as a separate commit.

Reply:
I think the explicit `return nil` is a better pattern for that helper -- sorry I didn't see it the first time :grinning: